### PR TITLE
[WIP]Add plugin for mapping tolerations to RBAC and SAs

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
@deads2k - First pass at mapping tolerations to RBAC

I did not register this plugin in the api server and there are some edge cases that I need to add.

### CODE REVIEW

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	PluginName       = "PodTolerations"
+	masterToleration = "node-role.kubernetes.io/master"
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: this is our plugin, name it for us.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	"io"
+
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	PluginName       = "PodTolerations"
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: this should be something int he openshift namespace.  `scheduling.openshift.io/RestrictPodTolerations`

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go (hunk 1 above)
Comment (deads2k) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: wrong line

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+func NewPodTolerationsPlugin() admission.Interface {
+	return &podTolerations{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+func (p *podTolerations) Validate(attr admission.Attributes, _ admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") && attr.GetKind().GroupKind() != coreapi.Kind("Pod") {
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: and subresource should be len zero

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	for _, toleration := range tolerations {
+		allow, err := p.checkPodTolerationAccess(attr, toleration)
+		if err != nil {
+			return err
+		}
+		if !allow {
+			return fmt.Errorf("requested user %v doesn't have permission to provide the given tolerations", attr.GetUserInfo().GetName())
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: you need to return API errors.  `NewForbidden` probably.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+func (p *podTolerations) validatePodTolerations(attr admission.Attributes, podSpec coreapi.PodSpec) error {
+	// if a pod has no tolerations, then no need to validate it's tolerations
+	if len(podSpec.Tolerations) == 0 {
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	for _, toleration := range tolerations {
+		allow, err := p.checkPodTolerationAccess(attr, toleration)
+		if err != nil {
+			return err
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: apierrors, so probalby NewInternalError

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+/*
+tolerations:
+- effect: NoSchedule
+  key: node-role.kubernetes.io/master
+  operator: Exists
+are converted to
+
+apiGroup: toleration.scheduling.openshift.io
+Resource: node-role.kubernetes.io/master
+resourceName: <empty_string>/NoSchedule
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: Maybe `<effect>::<value>`?  slashes aren't allowed in names and we don't want something unexpected to break.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+// checkPodTolerationAccess checks if pods are allowed to have a particular toleration. In this case, we're focusing
+// on master tolerations. For example,
+/*
+tolerations:
+- effect: NoSchedule
+  key: node-role.kubernetes.io/master
+  operator: Exists
+are converted to
+
+apiGroup: toleration.scheduling.openshift.io
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: please write corresponding RBAC rules for default cluster roles.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+Resource: node-role.kubernetes.io/master
+resourceName: <empty_string>/NoSchedule
+verbs: ["Exists"]
+*/
+func (p *podTolerations) checkPodTolerationAccess(attr admission.Attributes, toleration coreapi.Toleration) (bool, error) {
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: please extract the "toleration to attributesrecord" from the authorizer call so you can unit test it separately.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	verb := toleration.Operator
+	tolerationKey := ""
+	// TODO: Add some more cases for blanket toleration etc.
+	if len(toleration.Key) == 0 {
+		tolerationKey = masterToleration
+	}
+	authzAttr := authorizer.AttributesRecord{
+		User:            attr.GetUserInfo(),
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: this looks wrong.  The user is supposed to be a serviceaccount only, right?

Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerations/admission.go]: remember that serviceaccounts have hardcoded groups,  you'll want to specify those too.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+		if !allow {
+			return admission.NewForbidden(attr, fmt.Errorf("requested user %v doesn't have permission to provide the given tolerations", podSpec.ServiceAccountName))
+		}
+	}
+	return nil
+}
+
+func (p *podTolerations) IsAuthorized(podSAName string, attr admission.Attributes, toleration coreapi.Toleration) bool{
+	var saInfo user.Info
+	saInfo = serviceaccount.UserInfo(attr.GetNamespace(), podSAName, "")
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: this one looks right and it appears to fill in all the groups for you.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func NewPodTolerationsPlugin() admission.Interface {
+	return &podTolerations{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+func (p *podTolerations) Validate(attr admission.Attributes, _ admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") && attr.GetKind().GroupKind() != coreapi.Kind("Pod") && len(attr.GetSubresource()) == 0 {
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: I'm not so great at parsing these.  Mind breaking them apart?

```go
if attr.groupResource != thing{
    return
}
if attr.groupkind != thing{
    return
}
if len() != 0{
    return
}
```

I think your conditional on len is backwards.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func (p *podTolerations) validatePodTolerations(attr admission.Attributes, podSpec coreapi.PodSpec) error {
+	// if a pod has no tolerations, then no need to validate it's tolerations
+	if len(podSpec.Tolerations) == 0 {
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	if len(podSpec.ServiceAccountName) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("service account is not provided. The restrict pod tolerations plugin is completely lost"))
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: more user focused.  `serviceaccounts are required on pods that want to use tolerations`

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	for _, toleration := range tolerations {
+		allow := p.IsAuthorized(podSpec.ServiceAccountName, attr, toleration)
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: aggregate all the failures so the user gets the whole list back at once.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+/*
+tolerations:
+- effect: NoSchedule
+  key: node-role.kubernetes.io/master
+  operator: Exists
+
+are converted to
+
+apiGroup: toleration.scheduling.openshift.io
+Resource: toleration
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: seems like this should be key

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 11 above)
Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Sure thing.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+		APIGroup:        "toleration.scheduling.openshift.io",
+		ResourceRequest: true,
+		Name:            tolerationKey,
+	}
+	return authzAttr
+}
+
+
+// checkPodTolerationAccess checks if pods are allowed to have a particular toleration. In this case, we're focusing
+// on master tolerations. For example,
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: the example is good, but please also list the generic transformation.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+tolerations:
+- effect: NoSchedule
+  key: node-role.kubernetes.io/master
+  operator: Exists
+
+are converted to
+
+apiGroup: toleration.scheduling.openshift.io
+Resource: toleration
+resourceName: <empty_string>/NoSchedule
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: I don't think you understand why the `value/effect` in the resource name makes sense.  Also, you can't split on slash here.  For splitting, you probably want the effect first, right?

Empty names match all names.  Specified names match their specification.  This construct makes it easy to allow someone to tolerate a particular key with any value or any effect.  It makes it possible to enumerate the key/effect tuple as a unit.

Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: The above example is wrong. Please look at https://github.com/openshift/origin/pull/23427/files#diff-b49487e41f10910366174f767ef67f90R99-R137

Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Ok, so, if someone wants to have toleration against any key with any value or any effect, they can use the https://github.com/openshift/origin/pull/23427/files#diff-b49487e41f10910366174f767ef67f90R121-R123

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+	if len(podSpec.Tolerations) == 0 {
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	if len(podSpec.ServiceAccountName) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("serviceaccounts are required on pods that want to use tolerations"))
+	}
+	for _, toleration := range tolerations {
+		allow := p.IsAuthorized(podSpec.ServiceAccountName, attr, toleration)
+		var errs []error
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: wrong scope. This needs to move out of the for loop.  Please add a multi-error unit test

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+		if !allow {
+			errs = append(errs, fmt.Errorf("serviceaccounts doesn't have permission to provide the given toleration %v on the pod", toleration.Key))
+		}
+		if len(errs) > 0 {
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: this should be below the for loop

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	tolerations := podSpec.Tolerations
+	if len(podSpec.ServiceAccountName) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("serviceaccounts are required on pods that want to use tolerations"))
+	}
+	for _, toleration := range tolerations {
+		allow := p.IsAuthorized(podSpec.ServiceAccountName, attr, toleration)
+		var errs []error
+		if !allow {
+			errs = append(errs, fmt.Errorf("serviceaccounts doesn't have permission to provide the given toleration %v on the pod", toleration.Key))
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: serviceaccount

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+			return admission.NewForbidden(attr, err)
+		}
+	}
+	return nil
+}
+
+func (p *podTolerations) IsAuthorized(podSAName string, attr admission.Attributes, toleration coreapi.Toleration) bool {
+	saInfo := serviceaccount.UserInfo(attr.GetNamespace(), podSAName, "")
+	// check for hard-coded serviceaccount groups
+	SAInNamespace := serviceaccount.UserInfo(attr.GetNamespace(), "system:serviceaccounts:"+attr.GetNamespace(), "")
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: this line and the one below it aren't necessary.  your `saInfo` is fully correct.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+	SAAcrossAllNamespace := serviceaccount.UserInfo("", "system:serviceaccounts", "")
+	return p.checkPodTolerationAccess(saInfo, attr, toleration) || p.checkPodTolerationAccess(SAInNamespace, attr, toleration) || p.checkPodTolerationAccess(SAAcrossAllNamespace, attr, toleration)
+}
+
+// buildTolerationsAuthorizationAttributes builds the tolerations attributes from the user
+func buildTolerationsAuthorizationAttributes(info user.Info, namespace string, toleration coreapi.Toleration) authorizer.Attributes {
+	verb := toleration.Operator
+	tolerationKey := toleration.Key
+	if len(toleration.Key) == 0 {
+		tolerationKey = masterToleration
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: I think this default should be `*`

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+	saInfo := serviceaccount.UserInfo(attr.GetNamespace(), podSAName, "")
+	// check for hard-coded serviceaccount groups
+	SAInNamespace := serviceaccount.UserInfo(attr.GetNamespace(), "system:serviceaccounts:"+attr.GetNamespace(), "")
+	SAAcrossAllNamespace := serviceaccount.UserInfo("", "system:serviceaccounts", "")
+	return p.checkPodTolerationAccess(saInfo, attr, toleration) || p.checkPodTolerationAccess(SAInNamespace, attr, toleration) || p.checkPodTolerationAccess(SAAcrossAllNamespace, attr, toleration)
+}
+
+// buildTolerationsAuthorizationAttributes builds the tolerations attributes from the user
+func buildTolerationsAuthorizationAttributes(info user.Info, namespace string, toleration coreapi.Toleration) authorizer.Attributes {
+	verb := toleration.Operator
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: tolower

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+A cluster role to access toleration resources should have rbac rules like this
+
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: tolerations-creator
+rules:
+- apiGroups: ["toleration.scheduling.openshift.io"]
+  resources: ["node-role.kubernetes.io/master", "node-role.kubernetes.io/worker",...]
+  verbs: ["Exists", "Equal"] This can also be verbs:["*"]
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: do not suggest `*`

**Comment by deads2k** [comment]
I'd like one e2e test that makes sure this is active please.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 22 above)
Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Yeah, I did that yesterday evening but did not push it. I am on it.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 21 above)
Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Fair enough.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 20 above)
Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: You mean, we just need saInfo? I thought you wanted hardcoded SA'es to be used as well.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 17 above)
Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Silly me, what was I thinking. Thanks for catching it :)

**Comment by ravisantoshgudimetla** [comment]
> I'd like one e2e test that makes sure this is active and working.

Started on it. Got distracted. Will get to it, once I am done with changes with other PRs.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+/*
+tolerations:
+- effect: NoSchedule
+  key: node-role.kubernetes.io/master
+  operator: Exists
+
+are converted to
+
+apiGroup: toleration.scheduling.openshift.io
+Resource: node-role.kubernetes.io/master
```
Comment (smarterclayton) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Is there a way to wildcard by label namespace?  There will be certain label sets that nodes shouldn't be allowed to label.  I suspect we're going to end up needing to have at minimum by namespace, by prefix, etc.

Can we use `namespace` in the rule for `node-role.kubernetes.io` or are you leveraging namespace?  I.e. then the label "name" == `resourceName`, the label namespace == `namespace`, and the resource == `tolerations`?

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 20 above)
Comment (deads2k) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: > You mean, we just need saInfo? I thought you wanted hardcoded SA'es to be used as well.

I clicked through. that method builds them out for you already.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 24 above)
Comment (deads2k) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Using the namespace as a real namespace means that an admin granted powers in one namespace can delegate those power to multiple service accounts.  If we override the meaning of namespace, then someone with the power to grant rbac roles in the those namespaces becomes unexpectedly  powerful.

How many different toleration keys do we really expect to have?

Comment (ravisantoshgudimetla) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: Limiting based on namespace is really problematic. @smarterclayton, is your concern more on the lines of we will end up having writing permissions for too many tolerations on the pod spec(across multiple repos)?

File: test/extended/authorization/tolerations_restriction.go [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = createSAWithTolerationsAccess(cs, f.Namespace.Name)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = createRoleBinding(cs, f.Namespace.Name)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = createPodWithSA(cs, f.Namespace.Name)
+			o.Expect(err).NotTo(o.HaveOccurred())
+		})
+	})
+	g.Describe("Pod without service account should fail", func() {
```
Comment (deads2k) [inline test/extended/authorization/tolerations_restriction.go]: This test doesn't do what you want. All pods have serviceaccounts enforced via admission.  You need to test and SA without permissions.

File: test/extended/authorization/tolerations_restriction.go [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+		g.It(fmt.Sprintf("should fail"), func() {
+			err = createPodWithOutSA(cs, f.Namespace.Name)
+			o.Expect(err).Should(o.HaveOccurred())
+		})
+	})
+
+})
+
+func createClusterRole(cs clientset.Interface, nsName string) error {
```
Comment (deads2k) [inline test/extended/authorization/tolerations_restriction.go]: misnamed.  This is a local role. Also, it appears to be hardcoded so please name it for what it does

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/kubeadmission/register.go [hunk 27]
```diff
@@ -62,6 +64,7 @@ var (
 		"security.openshift.io/SCCExecRestrictions",
 		"route.openshift.io/IngressAdmission",
 		"quota.openshift.io/ClusterResourceQuota",
+		"scheduling.openshift.io/RestrictPodTolerations",
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/kubeadmission/register.go]: nothing comes after quota

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+	"k8s.io/klog"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+	"strings"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	PluginName       = "scheduling.openshift.io/RestrictPodTolerations"
+	masterToleration = "node-role.kubernetes.io/master"
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: delete

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+func (p *podTolerations) Validate(attr admission.Attributes, _ admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") {
+		return nil
+	}
+	if attr.GetKind().GroupKind() != coreapi.Kind("Pod") {
+		return nil
+	}
+	if len(attr.GetSubresource()) != 0 {
+		return nil
+	}
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: You only need to run on create and update, right?

File: test/extended/authorization/tolerations_restriction.go (hunk 25 above)
Comment (ravisantoshgudimetla) [reply test/extended/authorization/tolerations_restriction.go]: True but the plugin also enforces the existence of SA, so this case is still valid, I'll add one more test with a SA without enough privileges but I wanted to make sure that the happy path works fine.

File: test/extended/authorization/tolerations_restriction.go (hunk 26 above)
Comment (ravisantoshgudimetla) [reply test/extended/authorization/tolerations_restriction.go]: Yup. Thanks for catching, will change it.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go (hunk 20 above)
Comment (deads2k) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: this comment is still outstanding.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go [hunk 30]
```diff
@@ -0,0 +1,168 @@
...
+// buildTolerationsAuthorizationAttributes builds the tolerations attributes from the user
+func buildTolerationsAuthorizationAttributes(info user.Info, namespace string, toleration coreapi.Toleration) authorizer.Attributes {
+	verb := strings.ToLower(string(toleration.Operator))
+	tolerationKey := toleration.Key
+	if len(toleration.Key) == 0 {
+		tolerationKey = "*"
+	}
+	authzAttr := authorizer.AttributesRecord{
+		User:            info,
+		Verb:            string(verb),
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go]: unnecessary cast

File: test/extended/authorization/tolerations_restriction.go [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = g.Describe("[Feature: Toleration restrictions] Tolerations restrictions should be functional", func() {
+	defer g.GinkgoRecover()
+	f := e2e.NewDefaultFramework("tolerations-restriction")
+	cs, err := e2e.LoadClientset()
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	g.Describe("Pod with service account", func() {
```
Comment (deads2k) [inline test/extended/authorization/tolerations_restriction.go]: grant permission to a named toleration and make sure that passes, another fails, and tolerating everything fails.

**Comment by derekwaynecarr** [comment]
If taints are recommended as a means to provide hard partitioning, I think it makes sense to enable an RBAC model for specifying specific tolerations on nodes.  Hard partitioning is desired for special resources like GPUs where you want to repel workloads from landing on that node due to cost (see: https://github.com/NVIDIA/gpu-operator#hard-partitioning).  Another potential example for hardware partitioning is around network security zones for particular instance types.  In that model, you may declare a node is in a particular DMZ and also want to require explicit RBAC to use that DMZ.

The problem with taints right now is it is intended for both hard and soft partitioning, but this PR doesn't give a great way to configure for both.  Is it possible we could either a) enable this behind a feature gate that says hard partitioning only, or b) identify a set of keys that require hard partitioning (role, dedicated hardware, network security zone)?

**Comment by derekwaynecarr** [comment]
In general, I am worried about the friction this will cause for anyone that was use tolerations for soft partitioning schemes.

**Pull request closed without merging**